src/irmin/tree.ml

```diff
@@ -381,48 +378,69 @@ module Make (P : Private.S) = struct
       let info = { hash; map; value; findv_cache } in
       { v; info }
 
+    let of_map m = of_v (Map m)
+    let of_hash repo k = of_v (Hash (repo, k))
+    let of_value ?updates repo v = of_v (Value (repo, v, updates))
+
+    let cached_hash t =
+      match (t.v, t.info.hash) with
+      | Hash (_, h), None ->
+          let h = Some h in
+          t.info.hash <- h;
+          h
+      | _, h -> h
+
+    let cached_map t =
+      match (t.v, t.info.map) with
+      | Map m, None ->
+          let m = Some m in
+          t.info.map <- m;
+          m
+      | _, m -> m
+
+    let cached_value t =
+      match (t.v, t.info.value) with
+      | Value (_, v, None), None ->
+          let v = Some v in
+          t.info.value <- v;
+          v
+      | _, v -> v
+
+    let info_is_empty i =
+      i.map = None && i.value = None && i.findv_cache = None && i.hash = None
+
     let clear_info_fields i =
       if not (info_is_empty i) then (
         i.value <- None;
         i.map <- None;
         i.hash <- None;
         i.findv_cache <- None)
 
-    let rec clear_elt ~max_depth depth v =
-      match v with
-      | `Contents (c, _) -> if depth + 1 > max_depth then Contents.clear c
-      | `Node t -> clear ~max_depth (depth + 1) t
+    let rec clear_elt ~max_depth depth _ = function
+      | `Contents (c, _) -> Contents.clear c
+      | `Node t -> clear_node ~max_depth (depth + 1) t
 
-    and clear_info ~max_depth ?v depth i =
-      let clear _ v = clear_elt ~max_depth depth v in
-      let () =
-        match v with
-        | Some (Value (_, _, Some um)) ->
+    and clear_node ~max_depth depth n =
+      if depth = max_depth then (
+        clear_info_fields n.info;
+        match n.v with Value (_, v, _) -> P.Node.Val.clear v | _ -> ())
```

Comment:
Why only call this in the `depth = max_depth` case?

Reply:
As I implemented it:
1. it keeps every caches before `max_depth,
2. discards caches at `max_depth`,
3. doesn't visit after `max_depth`.

This is not the behaviour of the order version, I've reverted this.